### ISSUE TITLE
Mypy misses a typing error

### ISSUE THREAD

**Issue by takis** [python/mypy#13594]
**Bug Report**

Mypy misses a typing error.

```python
class A:
    def __init__(self) -> None:
        self.x: int

a = A()
a.x = "e"

class B:
    def __init__(self, x, y: int) -> None:
        self.x: int = x
        self.y: int = y

b = B(x="e", y="f")

```

**To Reproduce**
Run mypy on the above source code.

**Expected Behavior**
The first class was copied from the docs:
https://mypy.readthedocs.io/en/stable/class_basics.html

An error message stating that  instance attribute `x` in class B has incompatible type "str", expected "int".

**Actual Behavior**

In class A, it reports a problem with the types as expected.
In class B, it does report a problem with parameter y, because it was specified
to be an integer in the __init__ method.
However, it does not report any problem regarding instance attribute `x` in class B.

```bash
example.py:7: error: Incompatible types in assignment (expression has type "str", variable has type "int")
example.py:16: error: Argument "y" to "B" has incompatible type "str"; expected "int"
Found 2 errors in 1 file (checked 1 source file)
```

I added instance attribute 'y' in class B to make sure mypy would check the types, and I have tried using `--check-untyped-defs` as noted in the docs:
https://mypy.readthedocs.io/en/stable/common_issues.html

**Your Environment**

- Mypy version used: 0.971
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10.1
- Operating system and version: Debian testing

**Comment by AlexWaygood** [comment]
This is expected behaviour.

When analysing your code, mypy analyses each function independently of the rest of your code. This is an important optimisation technique. It also considers all untyped arguments to be implicitly annotated with `Any`, in order to allow codebases to gradually adopt type hints.

That means that in your class `B`, you've effectively established the following constraints with your type annotation:

```python
class B:
    # Constraint 1: the signature of __init__:
    # x is considered by mypy to be implicitly `Any`, as it is unannotated; y is `int`; the return type is `None`
    def __init__(self, x, y: int) -> None:
        # Constraint 2: x must always be a type consistent with `int` when directly set on instances of this class
        self.x: int = x
        # Constraint 3: y must always be a type consistent with `int` when directly set on instances of this class
        self.y: int = y
```

When you call `B(x="e", y="f")`, _only the first constraint_ is considered by mypy. Mypy doesn't look at all at what happens _inside_ `B.__init__` when analysing this call (because of the fact that it analyses all functions independently, and it will have analysed what goes on inside the `__init__` method separately); it _only_ looks at the signature of `__init__`. And in your signature for `__init__`, `x` is unannotated, meaning mypy considers it to have an `Any` type, meaning this type error goes undetected.
